llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.cpp

```diff
@@ -10,10 +10,69 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/SandboxIR/Function.h"
 #include "llvm/SandboxIR/Instruction.h"
-
-using namespace llvm::sandboxir;
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Transforms/Vectorize/SandboxVectorizer/Passes/NullPass.h"
 
 namespace llvm::sandboxir {
+
+static cl::opt<bool>
+    PrintPassPipeline("sbvec-print-pass-pipeline", cl::init(false), cl::Hidden,
+                      cl::desc("Prints the pass pipeline and returns."));
+
+/// A magic string for the default pass pipeline.
+static const char *DefaultPipelineMagicStr = "*";
+
+static cl::opt<std::string> UserDefinedPassPipeline(
+    "sbvec-passes", cl::init(DefaultPipelineMagicStr), cl::Hidden,
+    cl::desc("Comma-separated list of vectorizer passes. If not set "
+             "we run the predefined pipeline."));
+
+static void registerAllRegionPasses(sandboxir::PassRegistry &PR) {
+  PR.registerPass(std::make_unique<sandboxir::NullPass>());
+}
+
+/// Adds to `RPM` a sequence of passes described by `Pipeline` from the `PR`
+/// pass registry.
+static void parseAndCreatePassPipeline(RegionPassManager &RPM, PassRegistry &PR,
```

Comment:
Is this function going to be used for non RegionPasses too? If not, should it be a member function of RPM?

Reply:
It could be a member of RPM. The moral equivalent of this function for `FunctionPassManager` is currently a member function of `PassRegistry`:
```
class PassRegistry {
  ...
  FunctionPassManager &parseAndCreatePassPipeline(StringRef Pipeline);
};
```

If we change `PassRegistry::parseAndCreatePassPipeline` to get the pass manager as an argument rather than creating it inside the function, I think we could easily make it generic to avoid duplicating the pipeline parsing code:
```
class PassRegistry {
  ...
  template <typename PassManagerType>
  void parseAndCreatePassPipeline(PassManagerType &PM, StringRef Pipeline);
};
```

What do you think?

Reply:
(and maybe change the `Create` part of the `parseAndCreatePassPipeline` name to something that better describes the new behavior)

Reply:
Hmm then perhaps we should deprecate the PassRegistry in favor of a PassManager that owns the passes?

Reply:
We still need something to hold the mapping between names and passes so we can dynamically create pass pipelines based on the flag value. Right now the PassRegistry is that something. What are you proposing exactly?

Reply:
I don't see why we need a parsePassPipeline() in more than one place. Is the PassRegistry's `parsePassPipeline()` still being used after this patch?

Reply:
> Is the PassRegistry's `parsePassPipeline()` still being used after this patch?

As it is, no. That's why I proposed templatizing it so we can create pipelines of sandbox IR function passes, region passes, etc with a single parsing function. In the future, if there is more than one sandbox IR function pass, we'll probably want to create arbitrary pipelines out of them too. But I can also just delete that member function if we think we aren't going to need it.

Reply:
Yeah templatizing it is fine, we can do it when needed.

> But I can also just delete that member function if we think we aren't going to need it.

That's what I was thinking of. It seems that we are not going to need it.

Also if the PassRegistry's role ends up being just a mapping from name to pass class, then we could use a .def file.

Reply:
How about this?